File: server/lib/server/bookings.ex
defmodule Server.Bookings do
  import Ecto.Query
  alias Server.Repo
  alias Server.Bookings.Booking

  def list_bookings(params \\ %{}) do
    Booking
    |> filter_by_date(params["where"])
    |> filter_by_search(params["q"])
    |> filter_by_status(params["statuses"])
    |> sort_bookings(params["sort"], params["order"])
    |> paginate(params["page"], params["pageSize"])
    |> Repo.all()
  end

  def get_booking(id) do
    case Repo.get(Booking, id) do
      nil -> {:error, :not_found}
      booking -> {:ok, booking}
    end
  end

  def create_booking(attrs \\ %{}) do
    %Booking{}
    |> Booking.changeset(attrs)
    |> Repo.insert()
  end

  def update_booking(%Booking{} = booking, attrs) do
    booking
    |> Booking.changeset(attrs)
    |> Repo.update()
  end

  def delete_booking(%Booking{} = booking) do
    Repo.delete(booking)
  end

  # Move filtering/sorting to database level
  defp filter_by_search(query, nil), do: query

  defp filter_by_search(query, search_term) do
    from b in query,
      join: u in assoc(b, :user),
      where: ilike(u.name, ^"%#{search_term}%")
  end

  defp paginate(query, nil), do: query

  defp paginate(query, page, page_size) do
    from q in query,
      limit: ^page_size,
      offset: ^((page - 1) * page_size)
  end

  defp filter_by_date(query, nil), do: query

  defp filter_by_date(query, %{"start" => start_date, "end" => end_date}) do
    from b in query,
      where:
        b.date >= ^Date.from_iso8601!(start_date) and
          b.date <= ^Date.from_iso8601!(end_date)
  end

  defp filter_by_status(query, nil), do: query

  defp filter_by_status(query, statuses) when is_list(statuses) do
    from b in query, where: b.status in ^statuses
  end

  defp sort_bookings(query, nil, _), do: query

  defp sort_bookings(query, field, order) when field in ["date", "time", "duration"] do
    from b in query, order_by: [{^sort_direction(order), ^String.to_atom(field)}]
  end

  defp sort_direction("desc"), do: :desc
  defp sort_direction(_), do: :asc
end

File: client/server/db/bookings.ts
import type { BookingStatus } from "~/types";
import type { PgSelect } from "drizzle-orm/pg-core";
import { bookingParticipants, bookings, users } from "./schema";
import { and, eq, desc, asc, ilike, aliasedTable, lte, gt } from "drizzle-orm";

const db = useDrizzle();

// Alias for the users table
const participantUsers = aliasedTable(users, "participant_users");

export const getAllBookings = async (params: {
  q?: string;
  where: "future" | "all";
  statuses?: BookingStatus[];
  sort?: "date" | "user_id";
  order?: "asc" | "desc";
  page?: number;
  pageSize?: number;
}) => {
  const { q, where, statuses, sort, order, page, pageSize } = params;

  let query = db
    .select({
      booking: {
        id: bookings.id,
        date: bookings.date,
        time: bookings.time,
        duration: bookings.duration,
        status: bookings.status,
      },
      owner: {
        id: users.id,
        name: users.name,
        email: users.email,
        avatar: users.avatar,
        status: users.status,
        created_at: users.created_at,
        updated_at: users.updated_at,
        deleted_at: users.deleted_at,
      },
      participant: {
        id: bookingParticipants.id,
        paid: bookingParticipants.paid,
      },
      participantUser: {
        id: participantUsers.id,
        name: participantUsers.name,
        avatar: participantUsers.avatar,
        email: participantUsers.email,
        status: participantUsers.status,
        created_at: participantUsers.created_at,
        updated_at: participantUsers.updated_at,
        deleted_at: participantUsers.deleted_at,
      },
    })
    .from(bookings)
    .where(gt(bookings.date, new Date()).if(where === "future"))
    .leftJoin(users, eq(bookings.user_id, users.id))
    .leftJoin(
      bookingParticipants,
      eq(bookings.id, bookingParticipants.booking_id),
    )
    .leftJoin(
      participantUsers,
      eq(bookingParticipants.user_id, participantUsers.id),
    )
    .$dynamic();

  //if (where === "future") {
  //  query = query.where(gt(bookings.date, new Date()));
  //}

  // Apply search filter
  if (q) {
    query = query.where(ilike(users.name, `%${q}%`));
  }

  console.log(statuses);

  if (statuses) {
    // Apply status filter
    const statusArray = Array.isArray(statuses) ? statuses : [statuses];
    if (statusArray.length > 0) {
      query = query.where(
        and(...statusArray.map((status) => eq(bookings.status, status!))),
      );
    }
  }

  // Apply sorting
  if (sort) {
    const orderFunc = order === "desc" ? desc : asc;
    if (sort === "date") {
      query = query.orderBy(orderFunc(bookings.date));
    } else if (sort === "user_id") {
      query = query.orderBy(orderFunc(users.name));
    }
  }

  // Apply pagination
  if (page !== undefined && pageSize !== undefined) {
    query = withPagination(query, page, pageSize);
  }

  const results = await query;

  const response = mapQueryResultToResponse(results);

  return response;
};

type QueryResult = {
  booking: {
    id: number;
    date: Date;
    time: string;
    duration: number;
    status: string;
  };
  owner: User | null;
  participant: {
    id: number;
    paid: boolean | null;
  } | null;
  participantUser: User | null;
};

function mapQueryResultToResponse(results: QueryResult[]) {
  const bookingsMap = new Map();

  results.forEach((row) => {
    if (!bookingsMap.has(row.booking.id)) {
      bookingsMap.set(row.booking.id, {
        ...row.booking,
        owner: row.owner || null,
        participants: [],
      });
    }

    if (row.participant && row.participantUser) {
      bookingsMap.get(row.booking.id).participants.push({
        id: row.participant.id,
        paid: row.participant.paid,
        user: row.participantUser,
      });
    }
  });

  return Array.from(bookingsMap.values());
}

export const createBooking = async (bookingInput: {
  date: Date | string;
  time: string;
  duration: number;
  status: string;
  user_id: string;
}) => {
  // Parse the data if it's a string
  const parsedDate =
    bookingInput.date instanceof Date
      ? bookingInput.date
      : new Date(bookingInput.date);

  if (isNaN(parsedDate.getTime())) {
    throw new Error("Invalid date provided");
  }

  const bookingData = {
    ...bookingInput,
    date: parsedDate,
  };

  const newBooking = await db
    .insert(tables.bookings)
    .values(bookingData)
    .returning();

  return newBooking;
};

// Helper function for pagination
function withPagination<T extends PgSelect>(
  qb: T,
  page: number = 1,
  pageSize: number = 10,
) {
  return qb.limit(pageSize).offset((page - 1) * pageSize);
}
